code/go/0chain.net/blobbercore/readmarker/worker.go

```diff
@@ -6,113 +6,115 @@ import (
 	"github.com/0chain/blobber/code/go/0chain.net/core/chain"
-	"github.com/0chain/blobber/code/go/0chain.net/core/common"
-	zLogger "github.com/0chain/blobber/code/go/0chain.net/core/logging"
+	"github.com/0chain/blobber/code/go/0chain.net/core/logging"
 	"github.com/0chain/blobber/code/go/0chain.net/core/node"
 	"github.com/0chain/blobber/code/go/0chain.net/core/transaction"
-	"gorm.io/gorm"
 
 	"go.uber.org/zap"
 )
 
-func RedeemReadMarker(ctx context.Context, db *gorm.DB, rme *ReadMarkerEntity) error {
-	// Check if error is "already redeemed" then return nil
-	zLogger.Logger.Info("Redeeming read marker", zap.Any("rm", rme.ReadMarker))
+func SetupWorkers(ctx context.Context) {
+	go startRedeemMarkers(ctx)
+}
 
-	params := make(map[string]string)
-	params["blobber"] = node.Self.ID
-	params["client"] = rme.ReadMarker.ClientID
+func redeemReadMarker(ctx context.Context, rmEntity *ReadMarkerEntity) (err error) {
+	logging.Logger.Info("Redeeming the read marker", zap.Any("rm", rmEntity.LatestRM))
 
-	tx, err := transaction.NewTransactionEntity()
-	if err != nil {
-		return common.NewErrorf("redeem_read_marker", "creating transaction: %v", err)
+	params := map[string]string{
+		"blobber": rmEntity.LatestRM.BlobberID,
+		"client":  rmEntity.LatestRM.ClientID,
 	}
 
-	rdRedeem := &ReadRedeem{
-		ReadMarker: rme.ReadMarker,
-	}
-	rdRedeemBytes, err := json.Marshal(rdRedeem)
+	latestRM := ReadMarker{BlobberID: rmEntity.LatestRM.BlobberID, ClientID: rmEntity.LatestRM.ClientID}
+	latestRMBytes, err := transaction.MakeSCRestAPICall(
+		transaction.STORAGE_CONTRACT_ADDRESS, "/latestreadmarker", params,
+		chain.GetServerChain())
+
 	if err != nil {
-		zLogger.Logger.Error("Error encoding SC input", zap.Error(err), zap.Any("scdata", rdRedeem))
-		return common.NewErrorf("redeem_read_marker", "encoding SC data: %v", err)
+		logging.Logger.Error("Error from sc rest api call", zap.Error(err))
+		return
+	} else if err = json.Unmarshal(latestRMBytes, &latestRM); err != nil {
+		logging.Logger.Error("Error from unmarshal of rm bytes", zap.Error(err))
+		return
+	} else if latestRM.ReadCounter > 0 && latestRM.ReadCounter >= rmEntity.LatestRM.ReadCounter {
+		logging.Logger.Info("updating the local state to match the block chain")
+		if err = SaveLatestReadMarker(ctx, &latestRM, latestRM.ReadCounter, false); err != nil {
+			return
+		}
+		rmEntity.LatestRM = &latestRM
+		if err = rmEntity.Sync(ctx); err != nil {
+			logging.Logger.Error("redeem RM loop -- error syncing RM state", zap.Error(err))
+			return
+		}
+		return // synced from blockchain, no redeeming needed
 	}
 
-	if err := tx.ExecuteSmartContract(transaction.STORAGE_CONTRACT_ADDRESS, transaction.READ_REDEEM,
-		string(rdRedeemBytes), 0); err != nil {
-		zLogger.Logger.Info("Failed submitting read redeem", zap.Error(err))
-		return common.NewErrorf("redeem_read_marker", "sending transaction: %v", err)
+	if latestRM.ReadCounter == rmEntity.LatestRM.ReadCounter {
```

Comment:
Isn't this case is covered above in line 42? unless the latestRM.ReadCounter is 0, but is it necessary to run sync for it?

Reply:
Yes in current code , absolutely no need to have it. It was previous logic and it I think it was done to skip redeem request if rmEntity.LatestRM.ReadCounter is zero.